File: 74-search-a-2d-matrix/search-a-2d-matrix.java
class Solution {
    public static int binarySearch(int[] A, int target){
        int low = 0;
        int high = A.length-1;
        while(low <= high){
            int mid = low + ((high-low)/2);
            if(target == A[mid]) return mid;
            else if(target < A[mid]) high = mid-1;
            else low = mid + 1;
        }
        return -1;
    }
    public boolean searchMatrix(int[][] matrix, int target) {
        int rowlow = 0;
        int rowhigh = matrix.length-1;
        int col = matrix[0].length;

        while(rowlow <= rowhigh){
            int rowmid = rowlow + ((rowhigh-rowlow)/2);

            if(target >= matrix[rowmid][0] && target <= matrix[rowmid][col-1]){
                //int index = binarySearch(matrix[rowmid],target);
                //if(index != -1) return true;
                int low = 0;
                int high = matrix[rowmid].length-1;
                while(low <= high){
                    int mid = low + ((high-low)/2);
                    if(target == matrix[rowmid][mid]) return true;
                    else if(target < matrix[rowmid][mid]) high = mid-1;
                    else low = mid + 1;
                }
                return false;
            }
            else if(target > matrix[rowmid][col-1])
                rowlow = rowmid +1;
            else 
                rowhigh = rowmid - 1;
        }
        return false;
        
    }
}

File: 74-search-a-2d-matrix/search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
         //S1-Init Req DSs
    // track row and column
    int rows = matrix.size();
    int cols = matrix[0].size();
    int row = 0;
    int col = cols-1;

    // S2-Start Travarsal from top right corner and cont. till search space empty
    while(row < rows && col >= 0 ) {
        // A)Compare cur with target
        if(matrix[row][col] == target) {//found return true/pos
            return true;
        }
        else if (matrix[row][col] < target) {//B)skip cur row (Think carefully)
            row++;

        }
        else {//C)skip cur column
            col--;
        }
    }

    // S3-return false if traversal completed
    return false;

        
    }
};